bank workload report data inconsistency error on dr auto sync scenario

## Bug Report

<!-- Thanks for your bug report! Don't worry if you can't fill out all the sections. -->

### What version of TiKV are you using?
[2025/02/24 18:18:07.779 +08:00] [INFO] [cmd.go:168] ["Start remote command"] [cmd="/tiup/deploy/tikv-*/bin/tikv-server -V"] [nodename=tikv1]
TiKV 
Release Version:   9.0.0-alpha
Edition:           Community
Git Commit Hash:   c0b93db62631b730bf648488c388af8a1f2fb397
Git Commit Branch: HEAD
UTC Build Time:    2025-02-20 11:09:38
Rust Version:      rustc 1.77.0-nightly (89e2160c4 2023-12-27)
Enable Features:   memory-engine pprof-fp jemalloc mem-profiling portable sse test-engine-kv-rocksdb test-engine-raft-raft-engine trace-async-tasks openssl-vendored
Profile:           dist_release

### What operating system and CPU are you using?
8c/16GB

### Steps to reproduce
1、inject dc delay
2、run bank and tpcc

### What did you expect?
no data inconsistency

### What did happened?
bank workload report data inconsistency error
`
[2025/02/24 19:14:40.686 +08:00] [ERROR] [run.go:482] ["[%s] ADMIN CHECK TABLE bank2_accounts fails: %v"] [client=bank2] [error="Error 8223 (HY000): data inconsistency in table: bank2_accounts, index: byBalance, handle: 6576, index-values:\"handle: 6576, values: [KindInt64 1336]\" != record-values:\"handle: 6576, values: [KindInt64 1397]\""] [stack="github.com/pingcap/endless/pkg/workload/bank.(*Bank2Client).Verify\n\t/home/jenkins/agent/workspace/endless-master-build/pkg/workload/bank/run.go:482\ngithub.com/pingcap/endless/pkg/workload/bank.(*Bank2Client).ContinueVerify\n\t/home/jenkins/agent/workspace/endless-master-build/pkg/workload/bank/run.go:342\ngithub.com/pingcap/endless/testcase/oltp/tiuptest/drautosync/workload.RunBank.func2\n\t/home/jenkins/agent/workspace/endless-master-build/testcase/oltp/tiuptest/drautosync/workload/run.go:89\ngithub.com/pingcap/endless/testcase/oltp/tiuptest/drautosync/workload.RunBank.GinkgoRecoverWithErr.func4\n\t/home/jenkins/agent/workspace/endless-master-build/testcase/hatest/utils/fastfail.go:22\ngolang.org/x/sync/errgroup.(*Group).Go.func1\n\t/go/pkg/mod/golang.org/x/sync@v0.8.0/errgroup/errgroup.go:78"]`

@cfzjywxk This issue is unrelated to `dr-auto sync`; it’s purely a transaction bug introduced by #18200.

A simple way to reproduce the issue is by running the bank workload on a TiUP cluster:
1. Start a TiUP playground instance without TiFlash:
```tiup playground nightly --tiflash=0```

2. Use this script(https://github.com/3pointer/simpleWorkload/blob/main/README.md) to run workload
`go run main -dsn "root:@tcp(127.0.0.1:4000)/test" -accounts 1000 -concurrency 100 -duration 30m`

And wait at most 1 minute(sometimes seconds) the workload reports failed:
```
[ERROR] [main.go:388] ["Balance verification failed"] [expected=2000000] [actual=2000957]
```
and then doing a manually `admin check table bank2_accounts;` reports error.
```
mysql> admin check table bank2_accounts;
ERROR 8223 (HY000): data inconsistency in table: bank2_accounts, index: byBalance, handle: 1650, index-values:"handle: 1650, values: [KindInt64 333]" != record-values:"handle: 1650, values: [KindInt64 292]"
```
Reverting #18200 eliminates the issue.

Please review the changes in #18200 thoroughly, as they appear to introduce this transaction inconsistency.

--- a/src/storage/mvcc/reader/reader.rs
+++ b/src/storage/mvcc/reader/reader.rs
@@ -122,18 +122,6 @@ impl<S: EngineSnapshot> SnapshotReader<S> {
     pub fn take_statistics(&mut self) -> Statistics {
         std::mem::take(&mut self.reader.statistics)
     }
-
-    #[inline(always)]
-    pub fn setup_with_hint_items<T>(&mut self, items: &mut [T], key_of: fn(&T) -> &Key) {
-        // enable scan mode if there are multiple items, so that we don't need to seek
-        // for every key.
-        if items.len() > 1 {
-            items.sort_by(|a, b| key_of(a).cmp(key_of(b)));
-            self.reader.scan_mode = Some(ScanMode::Forward);
-            self.reader
-                .set_range(Some(key_of(items.first().unwrap()).clone()), None);
-        }
-    }
 }
 
 pub struct MvccReader<S: EngineSnapshot> {

--- a/src/storage/txn/commands/acquire_pessimistic_lock.rs
+++ b/src/storage/txn/commands/acquire_pessimistic_lock.rs
@@ -86,12 +86,12 @@ impl<S: Snapshot, L: LockManager> WriteCommand<S, L> for AcquirePessimisticLock 
             ))));
         }
 
-        let (start_ts, ctx, mut keys) = (self.start_ts, self.ctx, self.keys);
+        let (start_ts, ctx, keys) = (self.start_ts, self.ctx, self.keys);
         let mut txn = MvccTxn::new(start_ts, context.concurrency_manager);
-
-        let mut snapshot_reader = SnapshotReader::new_with_ctx(start_ts, snapshot, &ctx);
-        snapshot_reader.setup_with_hint_items(&mut keys, |k| &k.0);
-        let mut reader = ReaderWithStats::new(snapshot_reader, context.statistics);
+        let mut reader = ReaderWithStats::new(
+            SnapshotReader::new_with_ctx(start_ts, snapshot, &ctx),
+            context.statistics,
+        );
 
         let total_keys = keys.len();
         let mut res = PessimisticLockResults::with_capacity(total_keys);

--- a/src/storage/txn/commands/check_secondary_locks.rs
+++ b/src/storage/txn/commands/check_secondary_locks.rs
@@ -141,7 +141,7 @@ fn check_status_from_lock<S: Snapshot>(
 }
 
 impl<S: Snapshot, L: LockManager> WriteCommand<S, L> for CheckSecondaryLocks {
-    fn process_write(mut self, snapshot: S, context: WriteContext<'_, L>) -> Result<WriteResult> {
+    fn process_write(self, snapshot: S, context: WriteContext<'_, L>) -> Result<WriteResult> {
         // It is not allowed for commit to overwrite a protected rollback. So we update
         // max_ts to prevent this case from happening.
         let region_id = self.ctx.get_region_id();
@@ -152,10 +152,10 @@ impl<S: Snapshot, L: LockManager> WriteCommand<S, L> for CheckSecondaryLocks {
             })?;
 
         let mut txn = MvccTxn::new(self.start_ts, context.concurrency_manager);
-
-        let mut snapshot_reader = SnapshotReader::new_with_ctx(self.start_ts, snapshot, &self.ctx);
-        snapshot_reader.setup_with_hint_items(&mut self.keys, |k| k);
-        let mut reader = ReaderWithStats::new(snapshot_reader, context.statistics);
+        let mut reader = ReaderWithStats::new(
+            SnapshotReader::new_with_ctx(self.start_ts, snapshot, &self.ctx),
+            context.statistics,
+        );
         let mut released_locks = ReleasedLocks::new();
         let mut result = SecondaryLocksStatus::Locked(Vec::new());
 

--- a/src/storage/txn/commands/commit.rs
+++ b/src/storage/txn/commands/commit.rs
@@ -55,15 +55,15 @@ impl<S: Snapshot, L: LockManager> WriteCommand<S, L> for Commit {
             }));
         }
         let mut txn = MvccTxn::new(self.lock_ts, context.concurrency_manager);
-        let mut keys = self.keys;
-        let mut snapshot_reader = SnapshotReader::new_with_ctx(self.lock_ts, snapshot, &self.ctx);
-        snapshot_reader.setup_with_hint_items(&mut keys, |k| k);
-        let mut reader = ReaderWithStats::new(snapshot_reader, context.statistics);
+        let mut reader = ReaderWithStats::new(
+            SnapshotReader::new_with_ctx(self.lock_ts, snapshot, &self.ctx),
+            context.statistics,
+        );
 
-        let rows = keys.len();
+        let rows = self.keys.len();
         // Pessimistic txn needs key_hashes to wake up waiters
         let mut released_locks = ReleasedLocks::new();
-        for k in keys {
+        for k in self.keys {
             released_locks.push(commit(&mut txn, &mut reader, k, self.commit_ts)?);
         }
 

--- a/src/storage/txn/commands/flush.rs
+++ b/src/storage/txn/commands/flush.rs
@@ -2,8 +2,8 @@
 
 use std::mem;
 
-// #[PerformanceCriticalPath]
 use kvproto::kvrpcpb::{AssertionLevel, ExtraOp, PrewriteRequestPessimisticAction};
+// #[PerformanceCriticalPath]
 use txn_types::{insert_old_value_if_resolved, Mutation, OldValues, TimeStamp, TxnExtra};
 
 use crate::storage::{
@@ -76,10 +76,10 @@ impl<S: Snapshot, L: LockManager> WriteCommand<S, L> for Flush {
         }
         let rows = self.mutations.len();
         let mut txn = MvccTxn::new(self.start_ts, context.concurrency_manager);
-
-        let mut snapshot_reader = SnapshotReader::new_with_ctx(self.start_ts, snapshot, &self.ctx);
-        snapshot_reader.setup_with_hint_items(&mut self.mutations, |m| m.key());
-        let mut reader = ReaderWithStats::new(snapshot_reader, context.statistics);
+        let mut reader = ReaderWithStats::new(
+            SnapshotReader::new_with_ctx(self.start_ts, snapshot, &self.ctx),
+            context.statistics,
+        );
         let mut old_values = Default::default();
 
         let res = self.flush(&mut txn, &mut reader, &mut old_values, context.extra_op);

--- a/src/storage/txn/commands/prewrite.rs
+++ b/src/storage/txn/commands/prewrite.rs
@@ -537,9 +537,10 @@ impl<K: PrewriteKind> Prewriter<K> {
         self.check_max_ts_synced(&snapshot)?;
 
         let mut txn = MvccTxn::new(self.start_ts, context.concurrency_manager);
-        let mut snapshot_reader = SnapshotReader::new_with_ctx(self.start_ts, snapshot, &self.ctx);
-        snapshot_reader.setup_with_hint_items(&mut self.mutations, |m| m.key());
-        let mut reader = ReaderWithStats::new(snapshot_reader, context.statistics);
+        let mut reader = ReaderWithStats::new(
+            SnapshotReader::new_with_ctx(self.start_ts, snapshot, &self.ctx),
+            context.statistics,
+        );
         // Set extra op here for getting the write record when check write conflict in
         // prewrite.
 
@@ -615,6 +616,7 @@ impl<K: PrewriteKind> Prewriter<K> {
 
         // If there are other errors, return other error prior to `AssertionFailed`.
         let mut assertion_failure = None;
+
         for m in mem::take(&mut self.mutations) {
             let pessimistic_action = m.pessimistic_action();
             let expected_for_update_ts = m.pessimistic_expected_for_update_ts();
@@ -883,7 +885,6 @@ trait MutationLock {
     fn pessimistic_action(&self) -> PrewriteRequestPessimisticAction;
     fn pessimistic_expected_for_update_ts(&self) -> Option<TimeStamp>;
     fn into_mutation(self) -> Mutation;
-    fn key(&self) -> &Key;
 }
 
 impl MutationLock for Mutation {
@@ -897,10 +898,6 @@ impl MutationLock for Mutation {
 
     fn into_mutation(self) -> Mutation {
         self
-    }
-
-    fn key(&self) -> &Key {
-        self.key()
     }
 }
 
@@ -928,10 +925,6 @@ impl MutationLock for PessimisticMutation {
 
     fn into_mutation(self) -> Mutation {
         self.mutation
-    }
-
-    fn key(&self) -> &Key {
-        self.mutation.key()
     }
 }
 
@@ -1207,7 +1200,7 @@ mod tests {
         .unwrap();
         let d = perf.delta();
         assert_eq!(1, statistic.write.seek);
-        assert_eq!(d.internal_delete_skipped_count, 40);
+        assert_eq!(d.internal_delete_skipped_count, 0);
     }
 
     #[test]

--- a/src/storage/txn/commands/resolve_lock.rs
+++ b/src/storage/txn/commands/resolve_lock.rs
@@ -80,13 +80,13 @@ impl CommandExt for ResolveLock {
 
 impl<S: Snapshot, L: LockManager> WriteCommand<S, L> for ResolveLock {
     fn process_write(mut self, snapshot: S, context: WriteContext<'_, L>) -> Result<WriteResult> {
-        let (ctx, txn_status, mut key_locks) = (self.ctx, self.txn_status, self.key_locks);
+        let (ctx, txn_status, key_locks) = (self.ctx, self.txn_status, self.key_locks);
 
         let mut txn = MvccTxn::new(TimeStamp::zero(), context.concurrency_manager);
-
-        let mut snapshot_reader = SnapshotReader::new_with_ctx(TimeStamp::zero(), snapshot, &ctx);
-        snapshot_reader.setup_with_hint_items(&mut key_locks, |k| &k.0);
-        let mut reader = ReaderWithStats::new(snapshot_reader, context.statistics);
+        let mut reader = ReaderWithStats::new(
+            SnapshotReader::new_with_ctx(TimeStamp::zero(), snapshot, &ctx),
+            context.statistics,
+        );
 
         let mut scan_key = self.scan_key.take();
         let rows = key_locks.len();

--- a/src/storage/txn/commands/rollback.rs
+++ b/src/storage/txn/commands/rollback.rs
@@ -48,12 +48,12 @@ impl CommandExt for Rollback {
 }
 
 impl<S: Snapshot, L: LockManager> WriteCommand<S, L> for Rollback {
-    fn process_write(mut self, snapshot: S, context: WriteContext<'_, L>) -> Result<WriteResult> {
+    fn process_write(self, snapshot: S, context: WriteContext<'_, L>) -> Result<WriteResult> {
         let mut txn = MvccTxn::new(self.start_ts, context.concurrency_manager);
-
-        let mut snapshot_reader = SnapshotReader::new_with_ctx(self.start_ts, snapshot, &self.ctx);
-        snapshot_reader.setup_with_hint_items(&mut self.keys, |k| k);
-        let mut reader = ReaderWithStats::new(snapshot_reader, context.statistics);
+        let mut reader = ReaderWithStats::new(
+            SnapshotReader::new_with_ctx(self.start_ts, snapshot, &self.ctx),
+            context.statistics,
+        );
 
         let rows = self.keys.len();
         let mut released_locks = ReleasedLocks::new();

